src/libponyc/pass/docgen.c

```diff
@@ -548,6 +548,81 @@ static void doc_type_params(docgen_t* docgen, ast_t* t_params,
     fprintf(docgen->type_file, "]");
 }
 
+// Write the expression to the current type file.
+void doc_expression(docgen_t* docgen, ast_t* ast)
+{
+  FILE* fp = docgen->type_file;
+  ast_t* child = ast_child(ast);
+
+  switch(ast_id(ast))
+    {
+    case TK_CALL:
+      while(child != NULL)
+      {
+        doc_expression(docgen, child);
+        child = ast_sibling(child);
+      }
+      fprintf(fp, "()");
+      break;
+
+    case TK_SEQ:
+      while(child != NULL)
+      {
+        doc_expression(docgen, child);
+        child = ast_sibling(child);
+        if (child != NULL) {
+          fprintf(fp, "; ");
+        }
+      }
+      break;
+
+    case TK_REFERENCE:
+      doc_expression(docgen, child);
+      break;
+
+    case TK_TYPE:
+    case TK_INTERFACE:
+    case TK_TRAIT:
+    case TK_PRIMITIVE:
+    case TK_STRUCT:
+    case TK_CLASS:
+    case TK_ACTOR:
+      doc_type(docgen, ast, false);
+      break;
+
+    case TK_DOT:
+      while(child != NULL)
+        {
+          doc_expression(docgen, child);
+          child = ast_sibling(child);
+          if (child != NULL) {
+            fprintf(fp, ".");
+          }
+        }
+      break;
+
+    case TK_NONE:
+      break;
+
+    case TK_RECOVER:
+      fprintf(fp, "recover");
+      while(child != NULL)
+        {
+          if (ast_id(child) != TK_NONE) {
```

Comment:
standard pony c style formatting for this is something like

`  if(docgen.index_file != NULL && docgen.home_file != NULL)
  {
`